modules/grpc/bigquery/bigquery-worker.cpp

```diff
@@ -58,16 +88,308 @@ DestinationWorker::deinit()
+
+  std::chrono::system_clock::time_point connect_timeout =
+    std::chrono::system_clock::now() + std::chrono::seconds(10);
+
+  if (!this->channel->WaitForConnected(connect_timeout))
+    return false;
+
+  this->connected = true;
+  return true;
+}
+
+void
+DestinationWorker::disconnect()
+{
+  if (!this->connected)
+    return;
+
+  if (!this->batch_writer->WritesDone())
+    msg_warning("Error closing BigQuery write stream, writes may have been unsuccessful",
+                log_pipe_location_tag((LogPipe *) this->super->super.owner));
+
+  grpc::Status status = this->batch_writer->Finish();
+  if (!status.ok() && status.error_code() != grpc::StatusCode::INVALID_ARGUMENT)
+    {
+      msg_warning("Error closing BigQuery stream", evt_tag_str("error", status.error_message().c_str()),
+                  evt_tag_str("details", status.error_details().c_str()),
+                  evt_tag_int("code", status.error_code()),
+                  log_pipe_location_tag((LogPipe *) this->super->super.owner));
+    }
+
+  grpc::ClientContext ctx;
+  google::cloud::bigquery::storage::v1::FinalizeWriteStreamRequest finalize_request;
+  google::cloud::bigquery::storage::v1::FinalizeWriteStreamResponse finalize_response;
+  finalize_request.set_name(write_stream.name());
+
+  status = this->stub->FinalizeWriteStream(&ctx, finalize_request, &finalize_response);
+  if (!status.ok())
+    {
+      msg_warning("Error finalizing BigQuery write stream", evt_tag_str("error", status.error_message().c_str()),
+                  evt_tag_str("details", status.error_details().c_str()),
+                  log_pipe_location_tag((LogPipe *) this->super->super.owner));
+    }
+
+  this->connected = false;
+}
+
+void
+DestinationWorker::prepare_batch()
+{
+  this->batch_size = 0;
+  this->current_batch = google::cloud::bigquery::storage::v1::AppendRowsRequest{};
+
+  this->current_batch.set_write_stream(write_stream.name());
+  this->current_batch.set_trace_id("syslog-ng-bigquery");
+  google::cloud::bigquery::storage::v1::AppendRowsRequest_ProtoData *proto_rows =
+    this->current_batch.mutable_proto_rows();
+  google::cloud::bigquery::storage::v1::ProtoSchema *schema = proto_rows->mutable_writer_schema();
+  this->get_owner()->schema_descriptor->CopyTo(schema->mutable_proto_descriptor());
+}
+
+bool
+DestinationWorker::insert_field(const google::protobuf::Reflection *reflection, const Field &field,
+                                LogMessage *msg, google::protobuf::Message *message)
+{
+  DestinationDriver *owner = this->get_owner();
+
+  ScratchBuffersMarker m;
+  GString *value = scratch_buffers_alloc_and_mark(&m);
+
+  /* TODO trivial, NULL, etc. */
+  LogTemplateEvalOptions options = {&owner->template_options, LTZ_SEND, this->super->super.seq_num, NULL, LM_VT_STRING};
+  LogMessageValueType type;
+  log_template_format_value_and_type(field.value, msg, &options, value, &type);
+
+  if (type == LM_VT_NULL)
```

Comment:
I believe we should check whether the field is `required` before deciding whether to just skip a field or report an error or maybe fill the required field with a default instead.